cpp/src/fil/common.cuh

```diff
@@ -173,6 +172,101 @@ struct predict_params : shmem_size_params {
   int num_blocks;
 };
 
+template <bool COLS_IN_SHMEM, int LEAF_ALGO, int N_ITEMS>
+struct KernelTemplateParameters {
+  static const bool cols_in_shmem = COLS_IN_SHMEM;
+  static const leaf_algo_t leaf_algo = (leaf_algo_t)LEAF_ALGO;
+  static const int n_items = N_ITEMS;
+};
+
+namespace dispatch {
+
+template <template <class> class Func, class KernelParams, typename... Args>
+void dispatch_on_n_items(predict_params& params, Args... args) {
+  if (params.n_items == KernelParams::n_items) {
+    Func<KernelParams>(params, args...);
```

Comment:
It looks nice, but I'm not sure whether using a constructor call as a placeholder for a function call is a good idea.

I actually liked the idea of calling a templated `run()` on an object better. It would also reduce `template <class> class Func` to `typename Func` and remove the need for passing `Args...` around.

I'm also wondering if it would be possible to generate such an object automatically from a more common function definition. Such an automatic generator could then use `Args...`.